### PR TITLE
landing worker: better error handling and presentation (bug 1686869)

### PR DESCRIPTION
**Pull request by zzzeid**
- show rejected hunks/files (bug 1649556)
- show revision # of failing patch (bug 1687300)
- link to source changeset (bug 1687307)
- refactor error handling when applying patch
- rearrange try/except statements
- move .rej files to a temporary directory before cleaning repo

### CODE REVIEW

File: landoapi/models/landing_job.py [hunk 1]
```diff
@@ -246,6 +253,7 @@ def serialize(self):
                 {"revision_id": "D{}".format(r), "diff_id": self.revision_to_diff_id[r]}
                 for r in self.revision_order
             ],
+            "error_breakdown": self.error_breakdown,
```
Comment (zzzeid) [inline landoapi/models/landing_job.py:256]: Does this need to be added to the Swagger definition? (should probably review the consistency of swagger at some point, against the actual DB)

File: landoapi/landing_worker.py [hunk 2]
```diff
@@ -227,88 +246,177 @@ def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
...
+                            "path": r[0],
+                            "url": f"{repo.pull_path}/file/{r[1].decode('utf-8')}/{r[0]}",
+                            "changeset_id": r[1].decode("utf-8"),
+                        }
+                        for r in failed_path_changesets
+                    ]
+                    breakdown["reject_paths"] = {}
+                    for r in reject_paths:
+                        reject = {"path": r}
+                        # TODO: this should be a proper Path object, not a string.
```
Comment (zzzeid) [inline landoapi/landing_worker.py:330]: This TODO is no longer relevant.
```suggestion
```

File: landoapi/landing_worker.py [hunk 3]
```diff
@@ -227,88 +246,177 @@ def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
...
-            del os.environ["AUTOLAND_REQUEST_USER"]
+
+            # Set the environment variable the `set_landing_system.py` hg
+            # extension will use to send the push user override.
+            os.environ["AUTOLAND_REQUEST_USER"] = job.requester_email
+
+            try:
+                hgrepo.push(repo.push_path, bookmark=repo.push_bookmark or None)
+            except TreeClosed:
+                # error during push?
```
Comment (zzzeid) [inline landoapi/landing_worker.py:384]:
```suggestion
```

File: landoapi/landing_worker.py [hunk 4]
```diff
@@ -227,88 +246,177 @@ def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
...
+                job.transition_status(
+                    LandingJobAction.DEFER,
+                    message=f"Tree {repo.tree} is closed - retrying later.",
+                    commit=True,
+                    db=db,
+                )
+                return False
+            except TreeApprovalRequired:
+                # error during push
```
Comment (zzzeid) [inline landoapi/landing_worker.py:393]:
```suggestion
```

File: landoapi/landing_worker.py [hunk 5]
```diff
@@ -227,88 +246,177 @@ def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
...
+                job.transition_status(
+                    LandingJobAction.DEFER,
+                    message=f"Tree {repo.tree} requires approval - retrying later.",
+                    commit=True,
+                    db=db,
+                )
+                return False
+            except LostPushRace:
+                # push error
```
Comment (zzzeid) [inline landoapi/landing_worker.py:402]:
```suggestion
```

File: landoapi/models/landing_job.py [hunk 6]
```diff
@@ -102,6 +102,13 @@ class LandingJob(Base):
     # Text describing errors when status != LANDED.
     error = db.Column(db.Text(), default="")
 
+    # Error details in a dictionary format, listing failed merges, etc...
+    # E.g. {
+    #    "failed_paths": [{"path": "...", "url": "..."}],
+    #     "reject_paths": [{"path": "...", "content": "..."]
```
Comment (cgsheeh) [inline landoapi/models/landing_job.py:108]: Nit; `"}]`

File: landoapi/landing_worker.py [hunk 7]
```diff
@@ -227,88 +246,177 @@ def run_job(self, job, repo, hgrepo, treestatus, patch_bucket):
...
-                message=f"An unexpected error occured while landing:\n{e}",
-                commit=True,
-                db=db,
+                    job.transition_status(
+                        LandingJobAction.FAIL, message=message, commit=True, db=db
+                    )
+                    self.notify_user_of_landing_failure(job)
+                    return True
+                except NoDiffStartLine:
+                    # error with patch format? Where does this occur?
```
Comment (cgsheeh) [inline landoapi/landing_worker.py:347]: Is this a dev comment?

**Review by cgsheeh — APPROVED** [review]
Looking forward to seeing this in action!

File: landoapi/landing_worker.py (hunk 7 above)
Comment (zzzeid) [reply landoapi/landing_worker.py:347]: Yes I meant to remove those, will be removing them in next commit.

Comment (zzzeid) [inline landoapi/landing_worker.py:347]:
```suggestion
```

File: landoapi/models/landing_job.py (hunk 6 above)
Comment (zzzeid) [reply landoapi/models/landing_job.py:108]: Nice catch.

Comment (zzzeid) [inline landoapi/models/landing_job.py:108]:
```suggestion
    #     "reject_paths": [{"path": "...", "content": "..."}]
```

**Pull request merged**